File: src/com/example/algorithm/sorting/MergeSort.java
package com.example.algorithm.sorting;

import java.util.Arrays;

public class MergeSort {


    private  void mergeSort(int[] arr, int start, int end) {

        if (start < end) {
            int m = start+(end - start) / 2;
            mergeSort(arr, start, m);
            mergeSort(arr, m+1, end);
            merge(arr, start, m, end);
        }

    }


    private  void merge(int[] arr, int start, int mid, int end) {

        int l1 = mid - start + 1;
        int l2 = end - mid;


        int[] arr1 = new int[l1];
        int[] arr2 = new int[l2];

        for (int i = 0; i < l1; i++) {
            arr1[i] = arr[start + i];
        }

        for (int j = 0; j < l2; j++) {
            arr2[j] = arr[mid + 1 + j];
        }

        int i = 0, j = 0;
        int k = start;

        while (i<l1 && j<l2){
            if (arr1[i]<arr2[j]){
                arr[k] =arr1[i];
                i++;
            }else{
                arr[k] =arr2[j];
                j++;
            }
            k++;
        }
        while (i <l1){
            arr[k] =arr1[i];
            i++;
            k++;

        }
        while (j<l2){
            arr[k]  = arr2[j];
            j++;
            k++;
        }
    }


    public static void main(String[] args) {
        int[] arr=  {1, 5, 2, 1};
//        int[] arr = { 12, 11, 13, 5, 6, 7 };
        System.out.println(Arrays.toString(arr));
        new MergeSort().mergeSort(arr, 0, arr.length-1);
        System.out.println(Arrays.toString(arr));
    }

}

File: src/com/example/algorithm/MergeSort.kt
package com.example.algorithm


fun main() {
    val arr = intArrayOf(3, 2, 6, 5, 9, 4, 1)
    arr.forEach { print("$it, ") }
    MergeSort().mergeSort(arr, 0, arr.size - 1)
    println()
    arr.forEach { print("$it, ") }
}

class MergeSort {

    fun mergeSort(arr: IntArray, s: Int, e: Int) {
        if (e > s) {
            val m = s+(e - s) / 2
            mergeSort(arr, s, m)
            mergeSort(arr, m + 1, e)
            merge(arr, s, m, e)
        }
    }

    private fun merge(arr: IntArray, s: Int, m: Int, e: Int) {

        val l1 = m - s + 1
        val l2 = e - m

        val a1 = IntArray(l1)
        val a2 = IntArray(l2)

        for (i in a1.indices)
            a1[i] = arr[s + i]

        for (j in a2.indices)
            a2[j] = arr[m + 1 + j]


        var i = 0
        var j = 0
        var k = s

        while (i < l1 && j < l2) {
            if (a1[i] < a2[j]) {
                arr[k] = a1[i]
                i++
            } else {
                arr[k] = a2[j]
                j++
            }
            k++
        }

        while (i < l1) {
            arr[k] = a1[i]
            i++
            k++
        }
        while (j < l2) {
            arr[k] = a2[j]
            j++
            k++
        }
    }


}